modules/features2d/include/opencv2/features2d.hpp

```diff
@@ -1580,6 +1587,104 @@ class CV_EXPORTS_W BOWImgDescriptorExtractor
 
 //! @} features2d_category
 
+/****************************************************************************************\
+*                             Approximate Nearest Neighbors                              *
+\****************************************************************************************/
+
+//! @addtogroup features2d_annoy
+//! @{
+
+class CV_EXPORTS_W ANNIndex
+{
+public:
+    /** @brief Metrics used to calculate the distance between two feature vectors.
+     */
+    enum Distance
+    {
+        DIST_EUCLIDEAN,
+        DIST_MANHATTAN,
+        DIST_ANGULAR,
+        DIST_HAMMING,
+        DIST_DOTPRODUCT
+    };
+
+    virtual ~ANNIndex() = default;
+
+    /** @brief Add feature vectors to index.
+     *
+     * @param features Matrix containing the feature vectors to index. The size of the matrix is
+        num_features x feature_dimension.
+     */
+    CV_WRAP virtual void addItems(InputArray features) = 0;
+
+    /** @brief Build the index.
+     *
+     *  @param trees Number of trees in the index. If not provided, the number is determined automatically
+     *  in a way that at most 2x as much memory as the features vectors take is used.
+     */
+    CV_WRAP virtual void build(int trees=-1) = 0;
+
+    /** @brief Performs a K-nearest neighbor search for given query vector(s) using the index.
+     *
+     *  @param query The query vector(s).
+     *  @param indices Matrix that contains the indices of the K-nearest neighbors found. It must have
+        at least knn size.
+     *  @param dists Matrix that contains the distances to the K-nearest neighbors found. It must have
+        at least knn size.
```

Comment:
The description does not match implementation.
1. The both outputs are optional.
2. The output is re-created, if it has unexpected size or type.